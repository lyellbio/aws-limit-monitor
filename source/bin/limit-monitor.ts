#!/usr/bin/env node
/*****************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import * as cdk from '@aws-cdk/core';
import { LimitMonitorSpokeStack } from '../lib/limit-monitor-spoke-stack';
import { LimitMonitorStack } from '../lib/limit-monitor-stack';
import { ServiceQuotasChecksStack } from '../lib/service-quotas-checks-stack';

const SOLUTION_VERSION = process.env['DIST_VERSION'];
const SOLUTION_NAME = process.env['SOLUTION_NAME'];
const SOLUTION_ID = process.env['SOLUTION_ID'] || 'SO0005';
const SOLUTION_BUCKET = process.env['DIST_OUTPUT_BUCKET']; 
const SOLUTION_TMN = process.env['SOLUTION_TRADEMARKEDNAME'];
const SOLUTION_PROVIDER = 'AWS Solution Development';
const SOLUTION_TEMPLATE_BUCKET = process.env['DIST_TEMPLATE_BUCKET']

const app = new cdk.App();

export interface LimitMonitorStackProps extends cdk.StackProps {
  solutionId: string;
  solutionTradeMarkName: string | undefined;
  solutionProvider: string | undefined;
  solutionBucket: string | undefined;
  solutionTemplateBucket: string | undefined;
  solutionName: string | undefined;
  solutionVersion: string | undefined;
}

let limitMonitorStackProps: LimitMonitorStackProps = {
  solutionId: SOLUTION_ID,
  solutionTradeMarkName: SOLUTION_TMN,
  solutionProvider: SOLUTION_PROVIDER,
  solutionBucket: SOLUTION_BUCKET,
  solutionTemplateBucket: SOLUTION_TEMPLATE_BUCKET,
  solutionName: SOLUTION_NAME,
  solutionVersion: SOLUTION_VERSION,
  description: '(' + SOLUTION_ID + ') - ' + SOLUTION_NAME + ', version ' + SOLUTION_VERSION,
}
new LimitMonitorStack(app, 'limit-monitor', limitMonitorStackProps);

let serviceQuotasChecksStackProps: LimitMonitorStackProps = {
  solutionId: SOLUTION_ID,
  solutionTradeMarkName: SOLUTION_TMN,
  solutionProvider: SOLUTION_PROVIDER,
  solutionBucket: SOLUTION_BUCKET,
  solutionTemplateBucket: SOLUTION_TEMPLATE_BUCKET,
  solutionName: SOLUTION_NAME,
  solutionVersion: SOLUTION_VERSION,
  description: '(' + SOLUTION_ID + ') - ' + SOLUTION_NAME + ', version ' + SOLUTION_VERSION + ' - Spoke Template for vCPU limits',
}

new ServiceQuotasChecksStack(app, 'service-quotas-checks', serviceQuotasChecksStackProps);

let spokeTemplateProps: LimitMonitorStackProps = {
  solutionId: SOLUTION_ID,
  solutionTradeMarkName: SOLUTION_TMN,
  solutionProvider: SOLUTION_PROVIDER,
  solutionBucket: SOLUTION_BUCKET,
  solutionTemplateBucket: SOLUTION_TEMPLATE_BUCKET,
  solutionName: SOLUTION_NAME,
  solutionVersion: SOLUTION_VERSION,
  description: '(' + SOLUTION_ID + ') - ' + SOLUTION_NAME + ', version ' + SOLUTION_VERSION + ' - Spoke Template for vCPU limits',
}

new LimitMonitorSpokeStack(app, 'limit-monitor-spoke', spokeTemplateProps)
