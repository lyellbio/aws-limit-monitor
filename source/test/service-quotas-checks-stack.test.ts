#!/usr/bin/env node
/*****************************************************************************
 *  Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.   *
 *                                                                            *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may   *
 *  not use this file except in compliance with the License. A copy of the    *
 *  License is located at                                                     *
 *                                                                            *
 *      http://www.apache.org/licenses/LICENSE-2.0                            *
 *                                                                            *
 *  or in the 'license' file accompanying this file. This file is distributed *
 *  on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,        *
 *  express or implied. See the License for the specific language governing   *
 *  permissions and limitations under the License.                            *
 *****************************************************************************/

import { SynthUtils } from '@aws-cdk/assert';
import * as cdk from '@aws-cdk/core';
import { LimitMonitorStackProps } from '../bin/limit-monitor';
import serviceQuotasChecksStack = require('../lib/service-quotas-checks-stack');
import '@aws-cdk/assert/jest';


function getTestStack(): cdk.Stack {
  const app = new cdk.App();
  const envEU = { account: '111111111111', region: 'eu-west-1'};
  const props: LimitMonitorStackProps = {
    env: envEU,
    solutionId: 'SO0005',
    solutionBucket: 'solutions',
    solutionTemplateBucket: 'solutions-us-east-1',
    solutionName: 'limit-monitor',
    solutionVersion: 'v5.3.3',
    solutionProvider: 'AWS Solution Builders',
    solutionTradeMarkName: 'limit-monitor'
  }
  const stack = new serviceQuotasChecksStack.ServiceQuotasChecksStack(app, 'stack', props)
  stack.templateOptions.templateFormatVersion = "2010-09-09"
  return stack;
}

/*
 * SnapShot Testing for ServiceQuotasChecksStack.
 */
test('ServiceQuotasChecksStack snapshot test', () => {
  expect(SynthUtils.toCloudFormation(getTestStack())).toMatchSnapshot();
});