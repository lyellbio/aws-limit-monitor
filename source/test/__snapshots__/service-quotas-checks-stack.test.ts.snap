// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceQuotasChecksStack snapshot test 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Mappings": Object {
    "MetricsMap": Object {
      "Send-Data": Object {
        "SendAnonymousData": "Yes",
      },
    },
    "RefreshRate": Object {
      "CronSchedule": Object {
        "Default": "rate(5 minutes)",
      },
    },
    "SourceCode": Object {
      "General": Object {
        "KeyPrefix": "limit-monitor/v5.3.3",
        "S3Bucket": "solutions",
      },
    },
  },
  "Outputs": Object {
    "ServiceChecks": Object {
      "Description": "service limit checks monitored in the account",
      "Value": "vCPU",
    },
  },
  "Parameters": Object {
    "ThresholdPercentage": Object {
      "Default": 0.8,
      "Description": "This value is used to set the threshold for WARN messages. For example set this to 0.8 if the threshold percentage is 80%.",
      "Type": "Number",
    },
  },
  "Resources": Object {
    "LimitCheckSchedule": Object {
      "Properties": Object {
        "Description": "Limit Monitor Solution - Rule to perform limit checks",
        "ScheduleExpression": Object {
          "Fn::FindInMap": Array [
            "RefreshRate",
            "CronSchedule",
            "Default",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "LimitMonitorFunction",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "LimitCheckScheduleLambdaFunctionAwsEventsLambdaInvokePermission11533DD1E": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "LimitMonitorFunction",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "LimitCheckSchedule",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "LimitCheckScheduleLambdaFunctionServiceRole74B5A7C9": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "logs:CreateLogGroup",
                    "logs:CreateLogStream",
                    "logs:PutLogEvents",
                  ],
                  "Effect": "Allow",
                  "Resource": Object {
                    "Fn::Join": Array [
                      "",
                      Array [
                        "arn:",
                        Object {
                          "Ref": "AWS::Partition",
                        },
                        ":logs:",
                        Object {
                          "Ref": "AWS::Region",
                        },
                        ":",
                        Object {
                          "Ref": "AWS::AccountId",
                        },
                        ":log-group:/aws/lambda/*",
                      ],
                    ],
                  },
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "LambdaFunctionServiceRolePolicy",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LimitCheckScheduleLambdaFunctionServiceRoleDefaultPolicy757E0C96": Object {
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W11",
              "reason": "The actions servicequotas:GetServiceQuota, cloudwatch:GetMetricsData, events:PutEvents REQUIRE * resource.",
            },
            Object {
              "id": "W12",
              "reason": "Lambda needs the following minimum required permissions to send trace data to X-Ray.",
            },
          ],
        },
      },
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "servicequotas:GetServiceQuota",
                "servicequotas:GetAWSDefaultServiceQuota",
                "servicequotas:ListServiceQuotas",
                "cloudwatch:GetMetricData",
                "events:PutEvents",
                "ec2:DescribeRegions",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "logs:CreateLogGroup",
                "logs:CreateLogStream",
                "logs:PutLogEvents",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:",
                    Object {
                      "Ref": "AWS::Partition",
                    },
                    ":logs:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":log-group:/aws/lambda/*",
                  ],
                ],
              },
              "Sid": "default",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LimitCheckScheduleLambdaFunctionServiceRoleDefaultPolicy757E0C96",
        "Roles": Array [
          Object {
            "Ref": "LimitCheckScheduleLambdaFunctionServiceRole74B5A7C9",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LimitMonitorFunction": Object {
      "DependsOn": Array [
        "LimitCheckScheduleLambdaFunctionServiceRoleDefaultPolicy757E0C96",
        "LimitCheckScheduleLambdaFunctionServiceRole74B5A7C9",
      ],
      "Metadata": Object {
        "cfn_nag": Object {
          "rules_to_suppress": Array [
            Object {
              "id": "W89",
              "reason": "Not a valid use case to deploy in VPC",
            },
            Object {
              "id": "W92",
              "reason": "ReservedConcurrentExecutions not needed",
            },
          ],
        },
      },
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Join": Array [
              "-",
              Array [
                Object {
                  "Fn::FindInMap": Array [
                    "SourceCode",
                    "General",
                    "S3Bucket",
                  ],
                },
                Object {
                  "Ref": "AWS::Region",
                },
              ],
            ],
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "/",
              Array [
                Object {
                  "Fn::FindInMap": Array [
                    "SourceCode",
                    "General",
                    "KeyPrefix",
                  ],
                },
                "service-quotas-checks-service.zip",
              ],
            ],
          },
        },
        "Description": "This function checks for vCPU limits and sends notifiction on WARN and ERROR status",
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "LIMIT_THRESHOLD": Object {
              "Ref": "ThresholdPercentage",
            },
            "LOG_LEVEL": "INFO",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "LimitCheckScheduleLambdaFunctionServiceRole74B5A7C9",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
        "Timeout": 300,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
  },
}
`;
